File: C#/PaginationHelper.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace ConsoleApp6
{
    public class PagnationHelper<T>
    {
        public PagnationHelper(IList<T> collection, int itemsPerPage)
        {
            this.ItemsPerPage = itemsPerPage;
            this.ItemCount = collection.Count;
            Boolean even = this.ItemCount % this.ItemsPerPage == 0;
            this.PageCount = collection.Count / ItemsPerPage + (even ? 0 : 1);
        }


        private int itemsPerPage;
        private int itemCount;
        private int pageCount;


        public int ItemCount { get => itemCount; set => itemCount = value; }
        public int PageCount { get => pageCount; set => pageCount = value; }
        public int ItemsPerPage { get => itemsPerPage; set => itemsPerPage = value; }

        public int PageItemCount(int pageIndex)
        {
            Boolean even = this.ItemCount % this.ItemsPerPage == 0;
            //int totalPages = this.ItemCount / this.ItemsPerPage+(even?0:1);
            if (pageIndex + 1 > this.PageCount | pageIndex < 0)
            {
                return -1;
            }
            else if (pageIndex + 1 == this.PageCount & !even)
            {
                return this.ItemCount % this.ItemsPerPage;
            }
            else
            {
                return this.ItemsPerPage;
            }
        }

        public int PageIndex(int itemIndex)
        {

            Boolean even = this.ItemCount % this.ItemsPerPage == 0;
            if (itemIndex < 0 | itemIndex + 1 > this.ItemCount)
            {
                return -1;
            }
            else if (itemIndex + 1 > this.ItemsPerPage)
            {
                return (itemIndex + 1) / this.ItemsPerPage +
                    ((itemIndex + 1) % this.ItemsPerPage == 0 ? 0 : 1) - 1;
            }
            else
            {
                return 0;
            }
        }
    }
}

File: Python/pagination_helper.py
class PaginationHelper:

  # The constructor takes in an array of items and a integer indicating
  # how many items fit within a single page
  def __init__(self, collection, items_per_page):
    self.__collection=collection
    self.__items=items_per_page
      
  
  # returns the number of items within the entire collection
  def item_count(self):
    return len(self.__collection)
      
  
  # returns the number of pages
  def page_count(self):
    if self.item_count()%self.__items>0:
        adding=1
    else:
        adding=0
    return self.item_count()//self.__items+adding
      
    
  # returns the number of items on the current page. page_index is zero based
  # this method should return -1 for page_index values that are out of range
  def page_item_count(self,page_index):
    if page_index<0:
        return -1
    elif page_index<self.page_count()-1:
        return self.__items
    elif page_index==self.page_count()-1:
        return self.item_count()-page_index*self.__items
    else:
        return -1
  
  # determines what page an item is on. Zero based indexes.
  # this method should return -1 for item_index values that are out of range
  def page_index(self,item_index):
        if item_index+1%self.__items>0:
            adding=1
        else:
            adding=0
        if item_index<0:
            return -1
        elif item_index<self.item_count():
            return (item_index+1)//self.__items+adding-1
        else:
            return -1